### PR TITLE
fix Converter type indexing for Converters not supported by Extension

### PR DESCRIPTION
**Pull request by braingram**
tldr; #1561 introduced a subtle change to which Converters are ignored. stdatamodels includes an erroneous converter that is now not ignored. This PR fixes the asdf side of the issue (and drops the unused `gwcs` test dependency).

Prior to #1561 the types supported by a converter that doesn't support any tags that are supported by the extension were ignored. The modifications in that PR checked if the converter implemented 'select_tag' and if not ignored the types. However if the converter implemented 'select_tag' the converter types were indexed (as the converter could return None from 'select_tag' and defer to a new converter).

If a converter inherits from Converter it gains the select_tag method from the parent Converter class
(which provides documentation of the method but doesn't implement functionality that is otherwise covered by ConverterProxy). This creates an issue when indexing the converter tags/types if a converter is included in an extension that doesn't list support for the tags supported by the converter.

The changes included here removes the default 'select_tag' provided by Converter and adds a warning for any Converter subclasses that support more than 1 relevant tag in an extension and don't implement `select_tag`. This is a warning to provide backwards compatibility with extensions that relied on this bug.

As gwcs implements a few converters that subclass from Converter and support more than one tag, with the changes in this PR the gwcs extension will raise a warning (this is addressed in asdf-wcs-schemas PR: https://github.com/asdf-format/asdf-wcs-schemas/pull/46 and gwcs PR: https://github.com/spacetelescope/gwcs/pull/469 which splits the multiple tag versions across multiple extensions).

As gwcs is unused in the asdf tests, this PR removes gwcs as a test dependency and adds `asdf-astropy` (which was previously installed because of the `gwcs` dependency) for a few tests that rely on astropy objects.

### CODE REVIEW

File: asdf/extension/_converter.py [hunk 1]
```diff
@@ -211,9 +212,25 @@ def __init__(self, delegate, extension):
...
+            # find which class implements select_tag
+            for class_ in inspect.getmro(delegate.__class__):
+                if "select_tag" in class_.__dict__:
+                    if class_ is not Converter:
+                        # a non-Converter class implements select_tag so consider the types for this Converter
+                        break
+                    else:
+                        # this converter implements select_tag only because it inherits from Converter
+                        # don't index it's types
+                        return
```
[the remark this answers is not in the document]
Comment (braingram) [reply asdf/extension/_converter.py:233]: I think it should work to add an exception to the default `Converter.select_tag` method so that classes that inherit from `Converter` but are called with no tags have a more informative error (it was a puzzling one to untangle).

As far as supporting converters in extensions which support tags which aren't in the extension/manifest, it seems like a good thing to drop (although we should likely wait until stdatamodels is released which seems very likely to happen before asdf 3.0).

Dropping the gracious handling of the stdatamodels bug has the big benefit of allowing us to remove the more complicated part of this PR (the code using `inspect.getmro`). However, I do like that this catches the error (a tagless converter that doesn't re-implement `select_tag`) when the converter is proxied instead of when the converter gets used.

One thought that comes to mind would be to remove `select_tag` from the abstract `Converter`. This has one downside of removing the easy documentation of `select_tag` (is there another way we could document that method, ideally in a way that is runtime inspectable for someone testing out writing a converter in the repl?).

Let me make an attempt at updating this PR to:
- add an exception when the default `select_tag` is called with no tags
- drop the code supporting the ignoring of converters that support tags not mentioned in the manifest

@zacharyburnett do you know if there is a plan to release stdatamodels any time soon? The tldr; is that it relied on what is likely a bug in asdf which this PR fixes. The main branch of stdatamodels was fixed in https://github.com/spacetelescope/stdatamodels/pull/184 which has not yet made it into a release.

[the remark this answers is not in the document]
Comment (braingram) [reply asdf/extension/_converter.py]: It looks like removing the default `Converter.select_tag` did the trick.

I moved and slightly adapted the `select_tag` documentation to the `Converter` documentation (see: https://github.com/asdf-format/asdf/pull/1594/commits/18a3cc77b16d9ffa3c834054f38bc90bfcaa2edd#diff-14b06e3da444949d9a2c7f7b737b99b6c4112ede41f041a6d4f57b3627574d67R17).
I also changed the links to `select_tag` in the docs to link to `Converter` as that docstring now contains extra information about what is required of a `select_tag` method.

See the `Converter` docs here: https://asdf--1594.org.readthedocs.build/en/1594/api/asdf.extension.Converter.html#asdf.extension.Converter
and one of the modified links near the end of the section here: https://asdf--1594.org.readthedocs.build/en/1594/asdf/extending/converters.html#the-converter-interface

The devdeps failures appear unrelated and caused numpy 2.0 deprecating `maximum_sctype` https://github.com/numpy/numpy/pull/24154 which is used in astropy: https://github.com/astropy/astropy/blob/cd9dc7f818d97a64c2b5657a93fa3740a2270a38/astropy/nddata/bitmask.py#L22

[the remark this answers is not in the document]
Comment (braingram) [reply asdf/extension/_converter.py]: I re-added the above check (which causes the gwcs step and wcs converters to fail this check). I think the manifest in asdf-transform-schemas should be split into 2 versions and gwcs updated to register 2 extensions. I will put this PR into draft until those changes can be made because without them this causes many test failures due to warnings getting turned into errors.

**Comment by braingram** [comment]
I added another change into this PR to improve backwards compatibility. For a converter that supports more than 1 tag and inherits from `Converter` but no longer has a `select_tag` (because it was removed in this PR) now a warning (instead of an error) will be issued with a statement that this will become an error in a later version of asdf.

I'm still going to leave this as draft until the CI is repaired: https://github.com/asdf-format/asdf/pull/1629 but after that I think this should be fine to bring out of draft while the following is also worked on:

- asdf-wcs-schemas is updated to split the gwcs manifest: https://github.com/asdf-format/asdf-wcs-schemas/pull/45 https://github.com/asdf-format/asdf-wcs-schemas/pull/46
- gwcs is updated to use the new manifest(s): https://github.com/spacetelescope/gwcs/pull/469

After all of the above is sorted dev gwcs should work with no warnings and older gwcs versions will issue warnings (due to having converters that support multiple tags but don't define a select_tag).

**Comment by CagtayFabry** [comment]
> I added another change into this PR to improve backwards compatibility. For a converter that supports more than 1 tag and inherits from `Converter` but no longer has a `select_tag` (because it was removed in this PR) now a warning (instead of an error) will be issued with a statement that this will become an error in a later version of asdf.
> 
> I'm still going to leave this as draft until the CI is repaired: #1629 but after that I think this should be fine to bring out of draft while the following is also worked on:
> 
> * asdf-wcs-schemas is updated to split the gwcs manifest: [remove wcs schemas asdf-wcs-schemas#45](https://github.com/asdf-format/asdf-wcs-schemas/pull/45) [Split gwcs manifest asdf-wcs-schemas#46](https://github.com/asdf-format/asdf-wcs-schemas/pull/46)
> * gwcs is updated to use the new manifest(s): [register all gwcs manifests available from asdf-wcs-schemas spacetelescope/gwcs#469](https://github.com/spacetelescope/gwcs/pull/469)
> 
> After all of the above is sorted dev gwcs should work with no warnings and older gwcs versions will issue warnings (due to having converters that support multiple tags but don't define a select_tag).

I guess no longer relying on a default `select_tag` effectively also closes https://github.com/asdf-format/asdf/issues/1086 ?

(just leaving this here as xref for me, I hope you don't mind)

**Comment by braingram** [comment]
@CagtayFabry thanks for giving it a look!

This PR works towards but doesn't yet fix #1086. A converter that subclasses Converter will still end up selecting the oldest tag but with the changes in this PR a warning will be issued. Eventually we can turn this into an error but this would break backwards compatibility (and one of our key downstream packages [gwcs](https://github.com/spacetelescope/gwcs) which relies on this bug). I opened an issue to track turning the warning into an error. The change is currently milestoned for asdf 4.0 (which will hopefully come soon after 3.0 but almost certainly not for several months). I recently updated roadmaps for [3.0](https://github.com/asdf-format/asdf/wiki/3.0-Roadmap) and [beyond, including changes for 3.X.X](https://github.com/asdf-format/asdf/wiki/Roadmap) and any suggestions/feedback is welcome! It might be easiest to open new issues or comment on old ones to discuss changes. Maybe I should enable and learn how to use the 'discussions' :)

**Comment by braingram** [comment]
It looks like dkist also has a converter subclass that doesn't implement select_tag and support multiple tags. Here is the converter:
https://github.com/DKISTDC/dkist/blob/main/dkist/io/asdf/converters/dataset.py
and the manifest containing multiple dataset tag versions:
https://github.com/DKISTDC/dkist/blob/main/dkist/io/asdf/resources/manifests/dkist-0.9.0.yaml

**Comment by braingram** [comment]
I hope we have reached minimum yak hair.

This PR now also drop gwcs as a test dependency and adds asdf-astropy as a test dependency. gwcs was entirely unused and appears to only have been necessary because it installs asdf-astropy. Once this PR is done, some version of this PR: https://github.com/asdf-format/asdf/pull/1633 which cleans up our test dependencies should be able to remove asdf-astropy and astropy.

**Comment by braingram** [comment]
stdatamodels is failing because CRDS contains a known invalid ASDF file. This should be addressed in: https://github.com/spacetelescope/stdatamodels/pull/198

**Comment by braingram** [comment]
Closing and opening to trigger CI to run with newly merged PRs in asdf-wcs-schemas.

**Comment by braingram** [comment]
I suggest we hold off merging this until:
- https://github.com/spacetelescope/gwcs/pull/469 is merged (this should fix the gwcs downstream job)
- ideally gwcs is released with the above changes (so that the jwst downstream job doesn't fail due to gwcs warnings)

An alternative would be to ignore some of these warnings to allow the downstream CI to run (so we can see other errors) but this might hide other issues.

**Comment by braingram** [comment]
The jwst downstream test is still showing a `RuntimeError` for the gwcs extension.
https://github.com/asdf-format/asdf/actions/runs/6147488241/job/16688689417?pr=1594#step:10:13298
I did not expect this (given the released wcs schema, etc) and appears to be due to the gwcs `LabelMapperConverter` inheriting from `TransformConverterBase` which eventually inherits from `Converter`:
https://github.com/spacetelescope/gwcs/blob/83efcccc6ba1d65039eb84365223340a8e39344e/gwcs/converters/selector.py#L17C7-L17C27
combined with the use of `__bases__` (instead of `__mro__`) in this PR.

Taking the risk of pestering @eslavich for another review I switched this PR to use `__mro__` to hopefully improve the backwards compatibility with as many gwcs/asdf-wcs-schemas versions as possible.

With the above change we are back to just the expected unrelated jwst failures:
`UnboundLocalError: local variable 'n_slices' referenced before assignment`
caused by unreleased stcal changes (see: https://github.com/spacetelescope/jwst/pull/7874)

Sorry for revisiting this PR another time. It's hard to keep track of what are expected and unexpected failures :-/
